extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,309 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as st from '../../../src/style';
+import {Layout} from '../../../src/layout';
+import {assertHttpsUrl} from '../../../src/url';
+import {assert} from '../../../src/asserts';
+import {cidFor} from '../../../src/cid';
+import {getService} from '../../../src/service';
+import {vsyncFor} from '../../../src/vsync';
+import {xhrFor, createXhrRequest} from '../../../src/xhr';
+
+
+/**
+ * @typedef {{
+ *   elementId: string,
+ *   externalCID: string
+ * }}
+ */
+let GetRequestMetadata;
+
+/**
+ * @typedef {{
+ *   show_notification: boolean
+ * }}
+ */
+let GetResponseMetadata;
+
+/**
+ * @typedef {{
+ *
+ * }}
+ */
+let UserNotificationDefer;
+
+/**
+ *
+ */
+class AmpUserNotification extends AMP.BaseElement {
+
+  createdCallback() {
+    /** @private @const {!Window} */
+    this.win_ = this.getWin();
+
+    /** @private @const {!UserNotification} */
+    this.userNotification_ = getUserNotification_(this.win_);
+
+    /** @private @const {!Element} */
+    this.glasspane_ = this.userNotification_.glasspane;
+
+    // TODO: check with Dima if this is safe.
+    this.element.parentNode.removeChild(this.element);
+    this.glasspane_.appendChild(this.element);
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @private @const {!Vsync} */
+    this.vsync_ = vsyncFor(this.win_);
+
+    /** @private {boolean} */
+    this.activated_ = false;
+
+    assert(this.element.hasAttribute('id'), 'amp-user-notification should ' +
+        'have an id');
+    this.elementId_ = this.element.getAttribute('id');
+
+    /** @private @const {string} */
+    this.showIfHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-show-if-href'), this.element);
+
+    // assert these!
+    /** @private @const {string} */
+    this.dismissHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-dismiss-href'), this.element);
+
+    this.options_ = {
+      credentials: this.element.getAttribute('credentials')
+    };
+
+    /** @private @const {!function(*)} */
+    this.dialogResolve_;
```

Comment:
We should not do this. Closure now has @struct to signal that it isn't cool. The best practice changed in closure about 2 years ago or so.